### PR TITLE
(feat: M2-6790) Conditional logic improvements: Number select

### PR DESCRIPTION
**Pull request by moiskillnadne**
### 📝 Description
🔗 [Jira Ticket M2-6790](https://mindlogger.atlassian.net/browse/M2-6790)

### CODE REVIEW

File: src/features/PassSurvey/model/ConditionalLogicBuilder.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import { appletModel } from '~/entities/applet/';
+import { Condition } from '~/shared/api';
+
+type Item = appletModel.ItemRecord;
+
+export type ItemMapByName = Record<string, Item>;
+
+class ConditionalLogicBuilder {
+  private hiddenOrSkippedItemNames: Set<string> = new Set();
+  private hiddenOrSkippedItemIds: Set<string> = new Set();
```
Comment (BamMironov) [inline src/features/PassSurvey/model/ConditionalLogicBuilder.ts:12]: But how about creating an array like this:
```
private hiddenOrSkippedItems: Array<Item>
```

I understand the convenience of using Sets with primitives: it filters out duplicates automatically upon invoking "add". However, it populates the code with unnecessary lines likes this doing the same thing for the same objects:

this.hiddenOrSkippedItemNames.delete(item.name);
this.hiddenOrSkippedItemIds.delete(item.id);

File: src/widgets/Survey/model/hooks/useSurveyState.ts [hunk 2]
```diff
@@ -1,16 +1,17 @@
...
   const items = useMemo(() => activityProgress?.items ?? [], [activityProgress.items]);
 
   const visibleItems = items.filter((x) => !x.isHidden);
 
-  const processedItems = appletModel.conditionalLogicBuilder.process(visibleItems);
+  const processedItems = PassSurveyModel.conditionalLogicBuilder.process(visibleItems);
 
   const conditionallyHiddenItemIds =
-    appletModel.conditionalLogicBuilder.getConditionallyHiddenItemIds();
+    PassSurveyModel.conditionalLogicBuilder.getConditionallyHiddenItemIds();
```
Comment (BamMironov) [inline src/widgets/Survey/model/hooks/useSurveyState.ts:14]: I believe, at this point, we don't even need a separate method to return hidden items. We could either:
1) just figure out the hidden items right here, in this function, as we have all items (visibleItems) and  processedItems.
2) or make the `process` method return both hidden and visible items:
```
const availableItems = items.filter((x) => !x.isHidden);
const { visibleItems, hiddenItems } = PassSurveyModel.conditionalLogicBuilder.process(availableItems);
```

**Review by BamMironov — CHANGES_REQUESTED** [review]
Please consider changes below

File: src/features/PassSurvey/model/ConditionalLogicBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
+import { ConditionalLogicValidator } from './ConditionalLogicValidator';
+
+import { appletModel } from '~/entities/applet/';
+import { Condition } from '~/shared/api';
+
+type Item = appletModel.ItemRecord;
+
+export type ItemMapByName = Record<string, Item>;
+
+class ConditionalLogicBuilder {
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/features/PassSurvey/model/ConditionalLogicBuilder.ts:10]: Done

File: src/features/PassSurvey/model/ConditionalLogicBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  private hiddenOrSkippedItemNames: Set<string> = new Set();
+  private hiddenOrSkippedItemIds: Set<string> = new Set();
+
+  public process(items: Item[]): Item[] {
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/features/PassSurvey/model/ConditionalLogicBuilder.ts:14]: Done

File: src/features/PassSurvey/model/ConditionalLogicBuilder.ts (hunk 1 above)
Comment (moiskillnadne) [reply src/features/PassSurvey/model/ConditionalLogicBuilder.ts:12]: It doesn`t really help to reduce code. If we change a set to an array we should find the item each time when we want to manipulate it. 

So I don`t think that it improves readability or something. I suggest to keep it as is.

File: src/widgets/Survey/model/hooks/useSurveyState.ts (hunk 2 above)
Comment (moiskillnadne) [reply src/widgets/Survey/model/hooks/useSurveyState.ts:14]: Done

File: src/features/PassSurvey/model/ConditionalLogicBuilder.ts [hunk 5]
```diff
@@ -7,22 +7,32 @@ type Item = appletModel.ItemRecord;
 
 export type ItemMapByName = Record<string, Item>;
 
-class ConditionalLogicBuilder {
+type ProcessResult = {
+  visibleItems: Item[];
+  hiddenItemIds: Set<string>;
+};
+
+interface IConditionalLogicFilter {
```
Comment (BamMironov) [inline src/features/PassSurvey/model/ConditionalLogicBuilder.ts:15]: I suggest renaming it to `IFilter`and moving to the shared level

**Review by BamMironov — CHANGES_REQUESTED** [review]

File: src/features/PassSurvey/model/ConditionalLogicBuilder.ts (hunk 5 above)
Comment (moiskillnadne) [reply src/features/PassSurvey/model/ConditionalLogicBuilder.ts]: Done

File: src/shared/utils/types/filter.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+export interface IFilter<ItemProp, Result> {
+  filter: (items: ItemProp[]) => Result;
+}
```
Comment (BamMironov) [inline src/shared/utils/types/filter.ts:3]: I suggest changing it to this:
```
export interface IFilter<TItem, TResult = TItem[]> {
  filter(items: TItem[]): TResult;
}
```

1. Generic types are conventionally prefixed by "T"
2. `TResult` should be optional
3. Unless one wants to provide their own `TResult` type, the type should be the same `ItemProp[]`
4. I made it a class method, while yours is a property storing a function.

**Review by BamMironov — CHANGES_REQUESTED** [review]

File: src/shared/utils/types/filter.ts (hunk 6 above)
Comment (moiskillnadne) [reply src/shared/utils/types/filter.ts:3]: Done

**Review by BamMironov — APPROVED** [review]

**Pull request merged**